### ISSUE TITLE
Apply and test bundled tracers field in physics to dynamics remapper

### ISSUE THREAD

**Issue by jeff-cohere** [E3SM-Project/scream#758]
We have previously discussed how we might pack all advected tracer fields into a monster, multi-"component" "tracers" field for the dynamics package to advect. A detailed technical discussion started in #709, but I think there's enough detail that it deserves its own issue, so I've moved that conversation here.

## Approach

This task is closely coupled with the existing SCREAM physics atmosphere processes (p3, SHOC, RRTM), so it's important for us to lay out the work in stages. Most of this work involves the [PhysicsDynamicsRemapper class](https://github.com/E3SM-Project/scream/blob/master/components/scream/src/dynamics/homme/physics_dynamics_remapper.hpp). It would be much easier to wait until [Aaron's C++ P3 branch](https://github.com/E3SM-Project/scream/tree/aarondonahue/p3_interface_ad_cxx_redo) is merged, to make sure the P3 interface doesn't change during this task.

Currently, HOMME identifies tracers by querying the "extra data" in the headers of fields. As far as I can tell, this "extra data" is just a grab bag of "tags/labels/etc" that identify a field's properties. Unfortunately, we use the items "tag" and "label" for other things, so "extra data" is what we call it. If a field has an "extra data" called `"Is Tracer State"`, the physics-dynamics remapper interprets it as a tracer component. We've since decided to place tracer components into a field group to identify them as such. Switching to use field groups instead of this "extra data" is a good first step to take--it allows us to preserve other existing logic, but does require a little coordination with the P3 atmosphere process. 

Next, it remains to have the physics-dynamics remapper pack/unpack all of the fields identified as tracers into its own multi-component tracer field. This logic belongs in the following methods within that class:
* `do_remap_fwd`: the forward remapping process, which packs the tracers into a multi-component field. Specifically, we use the `itl` variable to indicate the tracer component index into which a tracer is packed, and the packing happens in `local_remap_fwd_2d` and `local_remap_fwd_3d` methods.
* `do_remap_bwd`: the backward remapping process, which unpacks the tracers from their multi-component field. Similar to `do_remap_fwd`, this method uses `itl` to extract tracers within the `remap_bwd_2d` and `remap_bwd_3d_impl` methods.

**NOTE: The `qv` field must be the first component in the multi-component tracers field.** 

Here, then, are all the subtasks associated with packing/unpacking tracer fields. Each gets its own PR:

1. Use an "advected" or "tracer" field group to identify tracer components (instead of using the unstructured field data)
2. Implement packing/unpacking of tracer fields in the forward and backward remapping methods of the physics-dynamics remapper class.

## Notes and Observations

Regarding "packing/unpacking": it seems like we might need a field of higher "rank" to store other fields that we want to combine. Such a "packed field" consists of any set of fields (identified by their "group", specified in the appropriateFieldTracking objects) that have been registered to be packed into a packed field with a specific field identifier. One might use a method like this one to register all fields in a specific group to be packed thus:

```
﻿﻿﻿﻿﻿   // Specify that all source fields belonging to the group identified by the
   // given case-insensitive string be packed into a "packed" field (of rank 2)
   // with the given target identifier. In the inverse mapping, these fields will
   // be unpacked from the source field into their target fields in the same
   // order in which they were packed.
   void register_packing(const ci_string& group_name,
                         const identifier_type& packed_tgt);
```

The individual fields being packed still need to be registered via `register_field` and bound to the remapper via `bound_field`. Other than that, they are treated as they would normally be treated by the remapper.

**Luca**: You would still need a call to register_field (and possibly bind_field) after that, though, right? I think what you suggest seems good. We need to find a way to enforce that all packing info is registered before any field is (I guess a simple assertion inside register_packing to check that no fields have been register yet, would do).

As for generality, I think we only have one use case in mind, so let's make this work for that case only, for now. However, thinking about it, the tracers might have a rank-3 packing (if I understand correctly your language). I say "might", cause I don't know what physics need. Homme stores qdp (tracers mass) with time slices, while q (tracers concentration) is "diagnostic". So their layouts are

qdp: (num_elems, num_q_time_levels, num_tracers, num_gp, num_gp, num_lev)
q  : (num_elems,                    num_tracers, num_gp, num_gp, num_lev)
Depending on what physics want, we might need to implement different things. If physics need qdp, we might have to do something that does packing in addition to time slice selection (which is what dynamics states already have to do).

I'm assuming physics only uses q, and can compute qdp if need be, but I'm not sure.

**Aaron**: Yes, you are correct. Currently physics only uses q. qdp is fully a dynamics variable. I think we can take advantage of this and keep things as simple up front. If, for some reason, someone decides they want physics to use qdp for something then that would likely be a mini-project of its own and they can cross that bridge then.

**Comment by bartgol** [comment]
The packed field is itself a Field. It only has a layout where the tag `Variable` or `Component` (or one of those) is added in somewhere. So I don't think we need a PackedFieldType.

**Comment by bartgol** [comment]
My biggest question is on the class in charge to setup the remapper (currently, that is `AtmosphereProcessGroup`: how would APG know which tgt field is the packed field to associate to all the unpacked src fields? Even better: how would APG know _when_ it is time to call `register_packing`?

I think the infrastructure you propose on the AbstractRemapper side (the added interface) is fine. I am trying to figure out how the outside world can use it.

**Comment by jeff-cohere** [comment]
That's a great question. Since we're using this to solve a unique problem at the moment, maybe we have logic in the APG that specifically designates a `tracers` field to be packed with all fields that belong to the `ADVECTED` group, if and only if it finds that dynamics is turned on? That's an ad-hoc solution, but it's nice and clear, and indicates what is going on.

To your edit: we could certainly associate the packed field's `FieldTracking` thing with a `PACKED` group to indicate that it's a target field. But what do you think about the simple ad-hoc approach above? I don't know if I'm missing something, of course. :-)

**Comment by bartgol** [comment]
I guess the ad-hoc solution would get the tracers problem solved. You could go ahead and do that, and add a small comment saying what could the strategy be for the general case if need be (e.g., add the packed field to the group "PACKED: <group name>", or some convention along that line). That way, if we later find out we have another use case for packing/unpacking, we may move to a general approach.

**Comment by jeff-cohere** [comment]
(Sorry for the pre-Thanksgiving bump! Just recording some thoughts before the week ends.)

Because we're relaxing the 1:1 relationship between remapped source and target fields, we have to decide how to accommodate the mapping of several source fields to a single "packed" field. One way to do this is to register all the source fields (in the order in which they appear in their group) with the single packed target field. And for the inverse mapping, you'd register all of the target fields against one "packed" source field. This assumes that one can have more than one instance of a field in the list of registered source fields, of course.

If this is reasonable, it might be easier for users to understand if we register/bind these fields for them when `register_packing` is called, instead of making them grapple with the one-to-many (or many-to-one) relationship themselves.

Alternatively, we can support a different (internal) registration process for packing/unpacking fields. Because fields themselves aren't stored within the AbstractRemapper, I'm a little worried that this might burden the implementor of a remapper subclass with understanding the detailed mechanics of packing/unpacking.

**Comment by bartgol** [comment]
I recently started to move some impl details of the grid class back to the base class. I figured it was saving me lots of similar implementation details (sometimes a bit tricky) in the derived classes. I think it may be reasonable to do something similar for the remappers, if that helps.

As for when we register the many-fields and packed-field, I don't have a strong preference. My main concern in this kind of decision is whether the classes _using_ the remappers have all the info they need to use the interfaces we add/change. So I'm ok having something like `register_packing (std::vector<field_type>& src_fields, field_type& tgt_field)` (and register_unpacking maybe as well). This would allow the remapper to remain a bit more agnostic to the impl details of Field (e.g., it would not need to know what a group/label in FieldTracking is). We can always manage the logic of how to figure out the correct many<->one matching in the encompassing classes (most likely only AtmProcGroup).

**Comment by AaronDonahue** [comment]
Good morning, I'm a bit late to this conversation, sorry about that.  I took the day before Thanksgiving off.  I just want to register an initial thought, apologies if it is a non-sequitur to this conversation.

Per "pack in the order the field is set", dynamics has one *hard* requirement for Q and Qdp.  That is that the water vapor mass mixing ratio, 'qv', be the first component in 'Q'.  We will need a way to enforce this in the packing setup.

**Comment by jeff-cohere** [comment]
That's fine--I think we can explicitly define the ordering wherever the field registration is done using @bartgol 's proposed method:

```
register_packing (std::vector<field_type>& src_fields, field_type& tgt_field)
```

We can stipulate that the ordering of the vector of `src_fields` is respected in the packing. This to me suggests that we shouldn't support a separate `register_unpacking` method, because the fact that we respect this order imposes a strong constraint on the inverse operation.

**Comment by bartgol** [comment]
@AaronDonahue The requirement you mention is only dictated by how dynamics uses Q and/or qdp, right? Meaning, you don't imply that there are other places in eam that _assume_ Q1 to be water vapor, right?

If so, the best solution would be to make the index of water vapor a configurable parameter in homme, so that it does something along the line of

```
   qv = elem(ie)%state%Q(:,:,:,qv_idx)
   [do something with qv]
```
I know preqx _does_ use qv, but I don't know where theta uses it. Perhaps in model initialization?

**Comment by AaronDonahue** [comment]
@bartgol , just to answer your question for posterity.  It's my understanding that physics won't have to worry about the indices of Q since the plan is to have physics explicitly define those fields outside of the Q structure.  So in short, I see only dynamics being concerned about the order of Q (and of course Qdp).  Surface coupling should also rely on fields as they are defined in physics.

So long story short.  I believe the answer is that only dynamics cares about qv being the first index of Q.

**Comment by PeterCaldwell** [comment]
Depends on #859 so putting back in "to-do" column. Once #859 is done, will be a very small task of just reshaping Q.

**Comment by bartgol** [comment]
I think bundled fields are correctly remapped. @tcclevenger correct me if I'm wrong, if field F1 has a "parent" F2 (the bundle of which F1 is a slice of), the PD remapper will _not_ remap F1, but I don't think it checks that this is true. So far, however, this is always the case, so we're good. I'm closing this for now.

**Comment by tcclevenger** [comment]
@bartgol Correct, if a field has a parent then we skip over it during remapping.
